doc/optimizers.md

```diff
@@ -1563,6 +1563,60 @@ optimizer.Optimize(f, coordinates);
  * [SGD in Wikipedia](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
  * [Differentiable separable functions](#differentiable-separable-functions)
 
+## NSGA2
+
+*An optimizer for aribitrary multi-objective functions.*
+
+NSGA2 (Non-dominated Sorting Genetic Algorithm - II) is a multi-objective optimization algorithm. The algorithm works by generating a candidate population from a fixed starting point. At each stage of optimization, a new population of children is generated. This new population along with its predecessor is sorted using non-domination as the metric. Following this, the population is further segregated into fronts. A new population is generated from these fronts having size equal to that of the starting population.
+
+#### Constructors
+
+ * `NSGA2()`
+ * `NSGA2(`_`populationSize, maxGenerations, crossoverProb, mutationProb, mutationStrength, epsilon, lowerBound, upperBound`_`)`
+
+
+#### Attributes
+
+| **type** | **name** | **description** | **default** |
+|----------|----------|-----------------|-------------|
+| `size_t` | **`populationSize`** | The number of candidates in the population. This should be at least 4 in size and a multiple of 4. | `100` |
+| `size_t` | **`maxGenerations`** | The maximum number of generations allowed for NSGA2. | `2000` |
+| `double` | **`crossoverProb`** | Probability that a crossover will occur. | `0.6` |
+| `double` | **`mutationProb`** | Probability that a weight will get mutated. | `0.3` |
+| `double` | **`mutationStrength`** | The range of mutation noise to be added. This range is between 0 and mutationStrength. | `0.001` |
+| `double` | **`epsilon`** | The final value of the objective function for termination. If set to negative value, tolerance is not considered. | `1e-6` |
+| `arma::vec` | **`lowerBound`** | Lower bound of the coordinates of the initial population. | `1` |
+| `arma::vec` | **`upperBound`** | Lower bound of the coordinates of the initial population. | `1` |
+
+
+Attributes of the optimizer may also be changed via the member methods
+`PopulationSize()`, `MaxGenerations()`, `CrossoverRate()`, `MutationProbability()`, `MutationStrength()`, `Epsilon()`, `LowerBound()` and `UpperBound()`.
+
+#### Examples:
+
+<details open>
+<summary>Click to collapse/expand example code.
+</summary>
+
+```c++
+SchafferFunctionN1<arma::mat> SCH;
+  arma::vec lowerBound("-1000 -1000");
+  arma::vec upperBound("1000 1000");
+  NSGA2 opt(20, 5000, 0.5, 0.5, 1e-3, 1e-6, lowerBound, upperBound);
+
+  arma::mat coords = SCH.GetInitialPoint();
+  auto objectives = SCH.GetObjectives();
```

Comment:
I can't seem to comment during a PR review on outdated code, so I can't continue this discussion directly: https://github.com/mlpack/ensmallen/pull/149#discussion_r370874927

Anyway, I'm wondering if actually it would be better to just define a `MultiObjectiveArbitraryFunction` type with the following API:

```
class MultiObjectiveArbitraryFunctionType
{
  // Evaluate all objectives.
  arma::Col<typename MatType::elem_type> Evaluate(const MatType& coordinates);
  // Get the number of objectives.
  size_t NumObjectives() const;
  // Evaluate an individual objective. (Is this useful?)
  typename MatType::elem_type Evaluate(const MatType& coordinates, const size_t objective);

};
```

It's not clear to me if it's more useful to have the batch `Evaluate()` or the individual `Evaluate()` (or both).  We could also create a nice utility wrapper that automatically wraps all of the objectives it's given, like `MultiObjectiveWrapper<Function1, Function2, Function3, ...>`.

Then, the code here could just be

```
arma::mat coords = SCH.GetInitialPoint();

std::vector<arma::mat> bestFront = opt.Optimize(objectives, coords);
```

Anyway, I think this is pretty similar to the design you and @zoq are already converging on---hopefully my thoughts are helpful here. :+1:  When we do settle on something, we'll have to update `function_types.md`.  (Also, can we open issues to add more multi-objective optimizers?  It'd be nice to have a list of multi-objective optimizers that didn't just have NSGA2 in it. :))

Reply:
Ryan, we came up with the tuple of objectives so that we could just pass a number of single-objective functions to the optimizer.

```
optimizer.Optimize(std::make_tuple(objectiveA, objectiveB, ... ), ...);
```
This way, we can use the preexisting `ArbitraryFunctionType`.

We can, however, have another utility wrapper with the current API that converts multiobjective functions (say, a new type named `ArbitraryMultiObjectiveFunctionType`) into a `std::tuple` of objectives along with their corresponding lower and upper bounds. What do you say?

Also, more MO optimizers would definitely be interesting!

Reply:
Sure, the `std::tuple` approach works for me; I guess I hadn't realized that there was consensus earlier.  If so I would have written a lot less. :smile:

Can I ask about the plan for documenting this approach?  Mostly I want to make sure it's visible and clear to users that they can do multi-objective optimization, and that they know how they can do it. :+1:

Reply:
Yes, Ryan, presently I have just added documentation for the NSGA2 optimizer to `doc/optimizers.md` file.  I think we should, however, illustrate the usage more verbosely for multi-objective optimisation framework. We can probably add documentation for multi-objective functions in `doc/function_types.md` too! 

What do you say? 😀

Reply:
I think that this is reasonable; I do think that we should add some documentation in `doc/function_types.md` also to cover this---I suspect that users figure out what they can optimize by looking there first. :)

Reply:
On it. :)

Reply:
Ryan, I have added documentation to `doc/function_types.md`, please have a look when you can! 😄